ffi/athcon/bindings/go/athcon.go

```diff
@@ -95,43 +98,40 @@ const (
 )
 
 type VM struct {
+	lib    uintptr
 	handle *C.struct_athcon_vm
 }
 
-func Load(filename string) (vm *VM, err error) {
-	cfilename := C.CString(filename)
-	loaderErr := C.enum_athcon_loader_error_code(C.ATHCON_LOADER_UNSPECIFIED_ERROR)
-	handle := C.athcon_load_and_create(cfilename, &loaderErr)
-	C.free(unsafe.Pointer(cfilename))
-
-	if loaderErr == C.ATHCON_LOADER_SUCCESS {
-		vm = &VM{handle}
-	} else {
-		errMsg := C.athcon_last_error_msg()
-		if errMsg != nil {
-			err = fmt.Errorf("ATHCON loading error: %s", C.GoString(errMsg))
-		} else {
-			err = fmt.Errorf("ATHCON loading error %d", int(loaderErr))
-		}
+func Load(path string) (*VM, error) {
+	handle, err := purego.Dlopen(path, purego.RTLD_NOW|purego.RTLD_GLOBAL)
+	if err != nil {
+		return nil, fmt.Errorf("loading library: %v", err)
 	}
 
-	return vm, err
+	filename := filepath.Base(path)
+	filename = strings.TrimSuffix(filename, filepath.Ext(filename))
+	vmName := strings.TrimPrefix(filename, "lib")
+
+	var athcon_create func() *C.struct_athcon_vm
+	purego.RegisterLibFunc(&athcon_create, handle, "athcon_create_"+vmName)
+	vmHandle := athcon_create()
+
+	if vmHandle == nil {
+		return nil, fmt.Errorf("failed to create VM")
+	}
+	return &VM{lib: handle, handle: vmHandle}, nil
```

Comment:
naming is quite confusing here. `handle` is a file handle on the dylib, and `vmHandle` is a pointer to the VM instance, right? we could probably make this a little clearer.

Reply:
Renamed `handle` to `libHandle`.